### PR TITLE
Add Addresses DAO

### PR DESCRIPTION
**Pull request by rostIvan**
```mysql
create table addresses
(
    id           int(11) unsigned not null auto_increment primary key,
    country      varchar(255)     null,
    city         varchar(255)     null,
    address      varchar(255)     null,
    zip          varchar(11)      null,
    latitude     float(10, 8)     null,
    longitude    float(11, 8)     null,
    warehouse_id int(11) unsigned null,
    associate_id int(11) unsigned null
);
```
| country | city | address | zip | latitude | longitude | warehouse_id | associate_id |
|---------|------|--------|-----|----------|-----------|---------------|--------------|
| Japan |  Tokyo | Kyobashi MID Bldg., 13-10, Kyobashi 2-chome, Chuo-ku |  153-0051 | 35.6850 | 139.7514 | null | 2|
| United States  | New York | 151 Pennington Street Brooklyn, NY 11236 |  10004 | 40.6943 | -73.9249 | 1 | null |

### CODE REVIEW

File: src/main/java/com/ita/if103java/ims/dao/impl/AddressDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+import java.util.Optional;
+
+@Repository
+public class AddressDaoImpl implements AddressDao {
+    private NamedParameterJdbcTemplate namedJdbcTemplate;
+    private AddressRowMapper mapper;
+
+    @Autowired
+    public AddressDaoImpl(DataSource dataSource, AddressRowMapper addressRowMapper) {
+        this.namedJdbcTemplate = new NamedParameterJdbcTemplate(dataSource);
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dao/impl/AddressDaoImpl.java]: Maybe create spring bean for NamedParameterJdbcTemplate?

File: src/main/java/com/ita/if103java/ims/dao/impl/AddressDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+            throw new CRUDException(crudErrorMessage, e);
+        }
+    }
+
+    public void updateSingleAddressRecord(String query, MapSqlParameterSource sqlParameterSource,
+                                          String notFoundMessage, String crudErrorMessage) {
+        try {
+            final int rowsAffected = namedJdbcTemplate.update(query, sqlParameterSource);
+            if (rowsAffected == 0) {
+                throw new EntityNotFoundException(notFoundMessage);
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dao/impl/AddressDaoImpl.java]: AddressNotFoundException

File: src/main/java/com/ita/if103java/ims/dao/impl/AddressDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+            throw new CRUDException(crudErrorMessage, e);
+        }
+    }
+
+    public Address findSingleAddressRecord(String query, MapSqlParameterSource sqlParameterSource,
+                                           String notFoundMessage, String crudErrorMessage) {
+        try {
+            return namedJdbcTemplate.queryForObject(query, sqlParameterSource, mapper);
+        } catch (EmptyResultDataAccessException e) {
+            throw new EntityNotFoundException(notFoundMessage);
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dao/impl/AddressDaoImpl.java]: AddressNotFoundException

File: src/main/java/com/ita/if103java/ims/dao/impl/AddressDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+                insert into addresses(associate_id, country, city, address, zip, latitude, longitude)
+                values (:associate_id, :country, :city, :address, :zip, :latitude, :longitude)
+            """;
+
+        public static final String SQL_UPDATE_WAREHOUSE_ADDRESS = """
+                update addresses
+                set country   = :country,
+                    city      = :city,
+                    address   = :address,
+                    zip       = :zip,
```
Comment (MarrMika) [inline src/main/java/com/ita/if103java/ims/dao/impl/AddressDaoImpl.java]: I think we should follow one code style(only JdbcTemplate or NamedParameterJdbcTemplate ).

**Comment by ikrasWork** [comment]
Please not use force push in git, you may brake something again :) , please use simple commits

**Pull request merged**